### PR TITLE
Fix logging for content downloader extensions

### PR DESCRIPTION
**Pull request by jw-msft**
## Bug Fixed

Fixes [Bug 43440951](https://microsoft.visualstudio.com/OS/_workitems/edit/43440951): deliveryoptimization_content_downloader.cpp trace not showing up in log

## Changes

- Add logging_manager to remove g_log_level hackiness, set log level at agent main at time of parsing cmdline args, and to get log level without having to modify fn signatures.
- Add OnDownloadBegin and OnDownloadEnd for begin and end of download thread execution
- Init logging in curl and DO content downloader OnDownloadBegin export (since Initialize is called in main thread and Download is called in download thread and log file paths are 1-1 with threads) and Uninit logging in OnDownloadEnd export
- Override OnDownloadBegin and OnDownloadEnd in StepsHandler impl to plumb through to the exports in the content downloader and leave the other content handlers with default, noop virtual method.
- Change some functions not used outside of CU to be static to file

## Result

After the changes, the following file is seen after doing apt pkg update:

```sh
/var/log/adu$ cat do-content-downloader.20230328-020410.log

2023-03-28T02:04:10.8268Z 315409[315415] [I] [do_download] ==== MULTI-LINE LOG BEGIN ====
Downloading File 'apt-manifest.json' from 'http://jewelden2--jewelden2.b.nlu.dl.adu.microsoft.com/eastus/jewelden2--jewelden2/272aeacb92594d51a422ed8d60f3a0c5/apt-manifest.json' to '/var/lib/adu/downloads/5935ab72-9122-4b7a-9405-a5b860b9bd87/apt-manifest.json'2023-03-28T02:04:10.8268Z 315409[315415] [I] [do_download] ==== MULTI-LINE LOG END ====

2023-03-28T02:04:11.5878Z 315409[315415] [I] Validating file hash [do_download]
2023-03-28T02:04:11.5880Z 315409[315415] [I] Download resultCode: 500, extendedCode: 0 [do_download]
====

```

### CODE REVIEW

**Comment by JeffMill** [comment]
the idea of a logging manager is good, but we already have a logging header which abstracts our logging layers (xlog, zlog).  Why not put it in there rather than adding a third layer of abstraction?

File: src/adu-shell/src/main.cpp [hunk 1]
```diff
@@ -145,8 +145,9 @@ int ParseLaunchArguments(const int argc, char** argv, ADUShell_LaunchArguments*
             launchArgs->logFile = optarg;
             break;
 
+        // clang-format off
```
Comment (JeffMill) [inline src/adu-shell/src/main.cpp:148]: why clang-format off?

File: src/agent/src/main.c [hunk 2]
```diff
@@ -294,8 +295,9 @@ int ParseLaunchArguments(const int argc, char** argv, ADUC_LaunchArguments* laun
             launchArgs->iotHubTracingEnabled = true;
             break;
 
+        // clang-format off
```
Comment (JeffMill) [inline src/agent/src/main.c:298]: here too - why clang-format off?

File: src/extensions/content_downloaders/curl_downloader/curl_content_downloader.EXPORTS.cpp [hunk 3]
```diff
@@ -28,6 +30,15 @@ ADUC_Result Download(
     unsigned int timeoutInSeconds,
     ADUC_DownloadProgressCallback downloadProgressCallback)
 {
+    // JIT init since Initialize is called by main thread and
```
Comment (JeffMill) [inline src/extensions/content_downloaders/curl_downloader/curl_content_downloader.EXPORTS.cpp:33]: not a fan of this.  I'd prefer to see a extension_Initialize() (and Uninitialize() ) function where this is called and uninitialized (as needed).  Relying on Download means there maybe be cases where logging cant occur in an extension

File: src/extensions/content_downloaders/deliveryoptimization_downloader/deliveryoptimization_content_downloader.EXPORTS.cpp [hunk 4]
```diff
@@ -94,6 +95,15 @@ ADUC_Result Download(
     unsigned int timeoutInSeconds,
     ADUC_DownloadProgressCallback downloadProgressCallback)
 {
+    // JIT init since Initialize is called by main thread and
```
Comment (JeffMill) [inline src/extensions/content_downloaders/deliveryoptimization_downloader/deliveryoptimization_content_downloader.EXPORTS.cpp:98]: see above

**Review by JeffMill — CHANGES_REQUESTED** [review]

File: src/agent/src/main.c (hunk 2 above)
Comment (jw-msft) [reply src/agent/src/main.c:298]: see above

File: src/adu-shell/src/main.cpp (hunk 1 above)
Comment (JeffMill) [reply src/adu-shell/src/main.cpp:148]: i'm fine with that style, and would look better than two clang-format comment lines.  alternately, remove the need for the braces (move stack var defn outside of switch scope)

File: src/extensions/content_downloaders/curl_downloader/curl_content_downloader.EXPORTS.cpp (hunk 3 above)
Comment (jw-msft) [reply src/extensions/content_downloaders/curl_downloader/curl_content_downloader.EXPORTS.cpp:33]: Agree. I think I could add a download_Init() and download_Uninit() that is called on download thread creation and destruction, respectively.

File: src/adu-shell/src/main.cpp (hunk 1 above)
Comment (jw-msft) [reply src/adu-shell/src/main.cpp]: removed clang-format off and reformatted.

File: src/extensions/content_downloaders/curl_downloader/curl_content_downloader.EXPORTS.cpp (hunk 3 above)
Comment (jw-msft) [reply src/extensions/content_downloaders/curl_downloader/curl_content_downloader.EXPORTS.cpp]: Added OnDownloadBegin and OnDownloadEnd.  It was missing a couple of traces earlier. It gets the following now in `/var/log/adu/do-content-downloader.*.log`:

```sh
2023-03-28T02:04:10.8268Z 315409[315415] [I] [do_download] ==== MULTI-LINE LOG BEGIN ====
Downloading File 'apt-manifest.json' from 'http://jewelden2--jewelden2.b.nlu.dl.adu.microsoft.com/eastus/jewelden2--jewelden2/272aeacb92594d51a422ed8d60f3a0c5/apt-manifest.json' to '/var/lib/adu/downloads/5935ab72-9122-4b7a-9405-a5b860b9bd87/apt-manifest.json'2023-03-28T02:04:10.8268Z 315409[315415] [I] [do_download] ==== MULTI-LINE LOG END ====

2023-03-28T02:04:11.5878Z 315409[315415] [I] Validating file hash [do_download]
2023-03-28T02:04:11.5880Z 315409[315415] [I] Download resultCode: 500, extendedCode: 0 [do_download]
```

File: scripts/error_code_generator_defs/result_codes.json [hunk 5]
```diff
@@ -652,6 +652,14 @@
               {
                 "name": "ADUC_ERC_CONTENT_DOWNLOADER_UNSUPPORTED_CONTRACT_VERSION",
                 "value": 13
+              },
+              {
+                "name": "ADUC_ERC_CONTENT_DOWNLOADER_ONDOWNLOADPROC_NOTIMP",
```
Comment (JeffMill) [inline scripts/error_code_generator_defs/result_codes.json:657]: ERC? Seems inconsistent, but it also looks like you're following what was in this file.

File: scripts/error_code_generator_defs/result_codes.json [hunk 6]
```diff
@@ -652,6 +652,14 @@
...
+                "value": 14
+              },
+              {
+                "name": "ADUC_ERC_CONTENT_DOWNLOADER_ONDOWNLOADENDPROC_NOTIMP",
```
Comment (JeffMill) [inline scripts/error_code_generator_defs/result_codes.json:661]: is there no ERRNO for NOTIMPLEMENTED?  I have a bug on this where our codes aren't generic, and so we end up with ONDOWNLOADPROC_NOTIMP, ONDOWNLOADPROCEND_NOTIMPL.  What happens if we add more functions that support NOTIMP, do we add FOO_NOTIMPL, BAR_NOTIMPL?

[Bug 43012743](https://microsoft.visualstudio.com/OS/_workitems/edit/43012743): Error codes are TOO content handler specific.

(marked as 2303, but I haven't seen progress)

File: src/extensions/content_downloaders/curl_downloader/curl_content_downloader.EXPORTS.cpp [hunk 7]
```diff
@@ -37,6 +39,26 @@ ADUC_Result Initialize(const char* initializeData)
     return { ADUC_GeneralResult_Success };
 }
 
+/**
+ * @brief Called on the worker context when execution is beginning.
+ *
+ */
+ADUC_Result OnDownloadBegin()
```
Comment (JeffMill) [inline src/extensions/content_downloaders/curl_downloader/curl_content_downloader.EXPORTS.cpp:46]: my preference (for now) would be an extension initialize/uninitialize.  

adding ondownloadbegin/end without ANY context is useless to an extension author, and appears only serving the purpose of allowing you to call logging init/uninit.

File: src/extensions/content_downloaders/deliveryoptimization_downloader/deliveryoptimization_content_downloader.EXPORTS.cpp [hunk 8]
```diff
@@ -77,6 +78,26 @@ ADUC_Result Initialize(const char* initializeData)
     return result;
 }
 
+/**
+ * @brief Called on the worker context when execution is beginning.
+ *
+ */
+ADUC_Result OnDownloadBegin()
```
Comment (JeffMill) [inline src/extensions/content_downloaders/deliveryoptimization_downloader/deliveryoptimization_content_downloader.EXPORTS.cpp:85]: as above.

File: src/extensions/extension_manager/inc/aduc/extension_manager.hpp [hunk 9]
```diff
@@ -17,7 +17,7 @@
 #include <aduc/types/download.h> // ADUC_DownloadProgressCallback
 #include <aduc/types/update_content.h> // ADUC_FileEntity
 
-#include <memory>
+#include <dlfcn.h>
```
Comment (JeffMill) [inline src/extensions/extension_manager/inc/aduc/extension_manager.hpp:20]: you'll want to build wiot.

**Review by JeffMill — CHANGES_REQUESTED** [review]

**Comment by jw-msft** [comment]
Closing this PR since there's a better/simpler fix that is significantly different. Will send it next week-ish.

**Pull request closed without merging**